src/mod_adlib.erl

```diff
@@ -25,10 +25,291 @@
+    sidejob_supervisor:spawn(
+            zotonic_sidejobs,
+            {?MODULE, import_all_async_task, [ Endpoint, ContextAsync ]}),
+    ok.
+
+%% @doc Async task for importing all Adlib docs, running as a zotonic_singular_job
+%% to prevent multiple imports running at the same time.
+-spec import_all_async_task( adlib_endpoint(), z:context() ) -> any().
+import_all_async_task(Endpoint, Context) ->
+    jobs:run(zotonic_singular_job, fun() -> import_all(Endpoint, Context) end).
+
+
+-spec import_all( Endpoint, Context ) -> {ok, Count} | {error, term()}
+    when Endpoint :: adlib_endpoint() | binary(),
+         Context :: z:context(),
+         Count :: non_neg_integer().
+import_all(Endpoint, Context) ->
+    import_since(Endpoint, undefined, Context).
+
+
+-spec import_since( Endpoint, Since, Context ) -> {ok, Count} | {error, term()}
+    when Endpoint :: adlib_endpoint() | binary(),
+         Since :: m_adlib_api:date(),
+         Context :: z:context(),
+         Count :: non_neg_integer().
+import_since(Name, Since, Context) when is_binary(Name) ->
+    case name_to_endpoint(Name, Context) of
+        {ok, E} -> import_since(E, Since, Context);
+        {error, _} = Error -> Error
+    end;
+import_since(#adlib_endpoint{} = Endpoint, Since, Context) ->
+    ?zInfo(
+        "Adlib: started import from endpoint '~s' since ~p",
+        [ Endpoint#adlib_endpoint.name, Since ],
+        Context),
+    import_all_loop(m_adlib_api:fetch_since(Endpoint, Since, Context), 0, Endpoint, Context).
+
+
+import_all_loop(done, Count, _Endpoint, _Context) ->
+    {ok, Count};
+import_all_loop({error, _} = Error, _Count, _Endpoint, _Context) ->
+    Error;
+import_all_loop({ok, {Recs, Cont}}, Count, Endpoint, Context) ->
+    Count1 = lists:foldl(
+        fun(#{ <<"@attributes">> := #{ <<"priref">> := PrirefAttr } }, Acc) ->
+            Priref = z_convert:to_integer(PrirefAttr),
+            case import_record(Endpoint, Priref, Context) of
+                {ok, Id} ->
+                    ?zInfo(
+                        "Adlib: import from endpoint '~s' priref '~p' to rsc id ~p",
+                        [ Endpoint#adlib_endpoint.name, Priref, Id ],
+                        Context),
+                    Acc+1;
+                {error, Reason} ->
+                    ?zWarning(
+                        "Adlib: import from endpoint '~s' priref '~p' error ~p",
+                        [ Endpoint#adlib_endpoint.name, Priref, Reason ],
+                        Context),
+                    Acc
+            end
+        end,
+        Count,
+        Recs),
+    case Cont of
+        {next, ContFun} ->
+            ?MODULE:import_all_loop(ContFun(), Count1, Endpoint, Context);
+        done ->
+            ?zInfo(
+                "Adlib: import from endpoint '~s' finished, imported ~p documents",
+                [ Endpoint#adlib_endpoint.name, Count1 ],
+                Context),
+            {ok, Count1}
+    end.
```

Comment:
How would a progress status work with these routines? Would being able to read out the queue length be best (might be, because the polling interval is then up to the frontend), or having a notifier at certain stages of the import (starting, per 10 percent, finished) - I guess I'd vote for the former.

Reply:
Maybe a notifier, yes. Then you can subscribe to it on a page.